File: framework/frontend/flutter/packages/k1s0_http/test/unit/api_error_test.dart
import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:k1s0_http/src/error/api_error.dart';
import 'package:k1s0_http/src/error/problem_details.dart';
import 'package:mocktail/mocktail.dart';

class MockDioException extends Mock implements DioException {}

class MockRequestOptions extends Mock implements RequestOptions {}

class MockResponse extends Mock implements Response<dynamic> {}

void main() {
  late MockRequestOptions mockRequestOptions;

  setUp(() {
    mockRequestOptions = MockRequestOptions();
    when(() => mockRequestOptions.extra).thenReturn({});
  });

  group('ApiErrorKind', () {
    test('contains all expected error kinds', () {
      expect(ApiErrorKind.values, contains(ApiErrorKind.validation));
      expect(ApiErrorKind.values, contains(ApiErrorKind.authentication));
      expect(ApiErrorKind.values, contains(ApiErrorKind.authorization));
      expect(ApiErrorKind.values, contains(ApiErrorKind.notFound));
      expect(ApiErrorKind.values, contains(ApiErrorKind.conflict));
      expect(ApiErrorKind.values, contains(ApiErrorKind.rateLimit));
      expect(ApiErrorKind.values, contains(ApiErrorKind.dependency));
      expect(ApiErrorKind.values, contains(ApiErrorKind.temporary));
      expect(ApiErrorKind.values, contains(ApiErrorKind.timeout));
      expect(ApiErrorKind.values, contains(ApiErrorKind.network));
      expect(ApiErrorKind.values, contains(ApiErrorKind.connection));
      expect(ApiErrorKind.values, contains(ApiErrorKind.cancelled));
      expect(ApiErrorKind.values, contains(ApiErrorKind.unknown));
    });
  });

  group('ApiError', () {
    test('creates with required fields', () {
      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
      );

      expect(error.kind, ApiErrorKind.validation);
      expect(error.message, 'Validation failed');
      expect(error.statusCode, isNull);
      expect(error.errorCode, isNull);
      expect(error.traceId, isNull);
      expect(error.problemDetails, isNull);
      expect(error.cause, isNull);
    });

    test('creates with all fields', () {
      const problemDetails = ProblemDetails(
        title: 'Validation Error',
        status: 400,
        errorCode: 'VALIDATION_ERROR',
      );
      final cause = Exception('Original error');
      final stackTrace = StackTrace.current;

      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
        statusCode: 400,
        errorCode: 'VALIDATION_ERROR',
        traceId: 'trace-123',
        problemDetails: problemDetails,
        cause: cause,
        stackTrace: stackTrace,
      );

      expect(error.statusCode, 400);
      expect(error.errorCode, 'VALIDATION_ERROR');
      expect(error.traceId, 'trace-123');
      expect(error.problemDetails, problemDetails);
      expect(error.cause, cause);
      expect(error.stackTrace, stackTrace);
    });

    test('requiresAuthentication returns true for authentication kind', () {
      final error = ApiError(
        kind: ApiErrorKind.authentication,
        message: 'Unauthorized',
      );

      expect(error.requiresAuthentication, true);
    });

    test('requiresAuthentication returns false for other kinds', () {
      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
      );

      expect(error.requiresAuthentication, false);
    });

    test('isRetryable returns true for retryable kinds', () {
      final retryableKinds = [
        ApiErrorKind.temporary,
        ApiErrorKind.dependency,
        ApiErrorKind.timeout,
        ApiErrorKind.network,
        ApiErrorKind.connection,
      ];

      for (final kind in retryableKinds) {
        final error = ApiError(kind: kind, message: 'Error');
        expect(error.isRetryable, true, reason: 'Expected $kind to be retryable');
      }
    });

    test('isRetryable returns false for non-retryable kinds', () {
      final nonRetryableKinds = [
        ApiErrorKind.validation,
        ApiErrorKind.authentication,
        ApiErrorKind.authorization,
        ApiErrorKind.notFound,
        ApiErrorKind.conflict,
        ApiErrorKind.rateLimit,
        ApiErrorKind.cancelled,
        ApiErrorKind.unknown,
      ];

      for (final kind in nonRetryableKinds) {
        final error = ApiError(kind: kind, message: 'Error');
        expect(error.isRetryable, false,
            reason: 'Expected $kind to not be retryable');
      }
    });

    test('fieldErrors returns errors from problemDetails', () {
      const problemDetails = ProblemDetails(
        title: 'Validation Error',
        status: 400,
        errorCode: 'VALIDATION_ERROR',
        errors: [
          FieldError(field: 'email', message: 'Invalid'),
        ],
      );

      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
        problemDetails: problemDetails,
      );

      expect(error.fieldErrors.length, 1);
      expect(error.fieldErrors[0].field, 'email');
    });

    test('fieldErrors returns empty list when problemDetails is null', () {
      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
      );

      expect(error.fieldErrors, isEmpty);
    });

    test('hasFieldErrors returns true when there are field errors', () {
      const problemDetails = ProblemDetails(
        title: 'Validation Error',
        status: 400,
        errorCode: 'VALIDATION_ERROR',
        errors: [
          FieldError(field: 'email', message: 'Invalid'),
        ],
      );

      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
        problemDetails: problemDetails,
      );

      expect(error.hasFieldErrors, true);
    });

    test('toString includes all relevant information', () {
      final error = ApiError(
        kind: ApiErrorKind.validation,
        message: 'Validation failed',
        statusCode: 400,
        errorCode: 'VALIDATION_ERROR',
        traceId: 'trace-123',
      );

      final str = error.toString();

      expect(str, contains('Validation failed'));
      expect(str, contains('VALIDATION_ERROR'));
      expect(str, contains('400'));
      expect(str, contains('trace-123'));
    });
  });

  group('ApiError.fromDioError', () {
    test('creates timeout error for connection timeout', () {
      final dioError = DioException(
        type: DioExceptionType.connectionTimeout,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError);

      expect(error.kind, ApiErrorKind.timeout);
      expect(error.message, 'Request timed out');
    });

    test('creates timeout error for send timeout', () {
      final dioError = DioException(
        type: DioExceptionType.sendTimeout,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError);

      expect(error.kind, ApiErrorKind.timeout);
    });

    test('creates timeout error for receive timeout', () {
      final dioError = DioException(
        type: DioExceptionType.receiveTimeout,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError);

      expect(error.kind, ApiErrorKind.timeout);
    });

    test('creates connection error for connection error', () {
      final dioError = DioException(
        type: DioExceptionType.connectionError,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError);

      expect(error.kind, ApiErrorKind.connection);
      expect(error.message, 'Connection failed');
    });

    test('creates cancelled error for cancel', () {
      final dioError = DioException(
        type: DioExceptionType.cancel,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError);

      expect(error.kind, ApiErrorKind.cancelled);
      expect(error.message, 'Request cancelled');
    });

    test('creates network error for bad certificate', () {
      final dioError = DioException(
        type: DioExceptionType.badCertificate,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError);

      expect(error.kind, ApiErrorKind.network);
      expect(error.message, 'Certificate verification failed');
    });

    test('preserves trace ID from request options', () {
      when(() => mockRequestOptions.extra)
          .thenReturn({'traceId': 'request-trace-123'});

      final dioError = DioException(
        type: DioExceptionType.connectionTimeout,
        requestOptions: mockRequestOptions,
      );

      final error = ApiError.fromDioError(dioError, 'trace-456');

      expect(error.traceId, 'trace-456');
    });
  });

  group('ApiError.fromResponse', () {
    late MockResponse mockResponse;

    setUp(() {
      mockResponse = MockResponse();
    });

    test('maps 400 to validation error', () {
      when(() => mockResponse.statusCode).thenReturn(400);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.validation);
    });

    test('maps 401 to authentication error', () {
      when(() => mockResponse.statusCode).thenReturn(401);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.authentication);
    });

    test('maps 403 to authorization error', () {
      when(() => mockResponse.statusCode).thenReturn(403);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.authorization);
    });

    test('maps 404 to not found error', () {
      when(() => mockResponse.statusCode).thenReturn(404);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.notFound);
    });

    test('maps 409 to conflict error', () {
      when(() => mockResponse.statusCode).thenReturn(409);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.conflict);
    });

    test('maps 429 to rate limit error', () {
      when(() => mockResponse.statusCode).thenReturn(429);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.rateLimit);
    });

    test('maps 502 to dependency error', () {
      when(() => mockResponse.statusCode).thenReturn(502);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.dependency);
    });

    test('maps 503 to dependency error', () {
      when(() => mockResponse.statusCode).thenReturn(503);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.dependency);
    });

    test('maps 504 to timeout error', () {
      when(() => mockResponse.statusCode).thenReturn(504);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.timeout);
    });

    test('maps other 5xx to temporary error', () {
      when(() => mockResponse.statusCode).thenReturn(500);
      when(() => mockResponse.data).thenReturn(null);

      final error = ApiError.fromResponse(mockResponse);

      expect(error.kind, ApiErrorKind.temporary);
    });

    test('parses ProblemDetails from response data', () {
      when(() => mockResponse.statusCode).thenReturn(400);
      when(() => mockResponse.data).thenReturn({
        'title': 'Validation Error',
        'status': 400,
        'error_code': 'VALIDATION_ERROR',
        'detail': 'The email field is invalid',
        'trace_id': 'trace-123',
      });

      final error = ApiError.fromResponse(mockResponse);

      expect(error.message, 'The email field is invalid');
      expect(error.errorCode, 'VALIDATION_ERROR');
      expect(error.traceId, 'trace-123');
      expect(error.problemDetails, isNotNull);
    });

    test('handles string response data', () {
      when(() => mockResponse.statusCode).thenReturn(500);
      when(() => mockResponse.data).thenReturn('Internal Server Error');

      final error = ApiError.fromResponse(mockResponse);

      expect(error.message, 'Internal Server Error');
    });

    test('uses default message when data is empty', () {
      when(() => mockResponse.statusCode).thenReturn(404);
      when(() => mockResponse.data).thenReturn('');

      final error = ApiError.fromResponse(mockResponse);

      expect(error.message, 'Resource not found');
    });
  });

  group('ApiErrorExtension', () {
    test('localizedMessage returns Japanese messages', () {
      expect(
        ApiError(kind: ApiErrorKind.validation, message: '')
            .localizedMessage
            .isNotEmpty,
        true,
      );
      expect(
        ApiError(kind: ApiErrorKind.authentication, message: '')
            .localizedMessage,
        contains('認証'),
      );
      expect(
        ApiError(kind: ApiErrorKind.timeout, message: '').localizedMessage,
        contains('タイムアウト'),
      );
      expect(
        ApiError(kind: ApiErrorKind.network, message: '').localizedMessage,
        contains('ネットワーク'),
      );
    });
  });
}

File: framework/frontend/react/packages/k1s0-api-client/tests/error/ApiError.test.ts
/**
 * ApiError のテスト
 */

import { describe, it, expect, vi } from 'vitest';
import { ApiError } from '../../src/error/ApiError';
import type { ApiErrorKind } from '../../src/error/types';

// Response モック
const createMockResponse = (
  body: unknown,
  options: { status: number; headers?: Record<string, string> }
) => {
  const headers = new Map(Object.entries(options.headers ?? {}));
  return {
    ok: options.status >= 200 && options.status < 300,
    status: options.status,
    headers: {
      get: (name: string) => headers.get(name.toLowerCase()) ?? null,
    },
    json: vi.fn().mockResolvedValue(body),
    text: vi.fn().mockResolvedValue(JSON.stringify(body)),
  } as unknown as Response;
};

describe('ApiError', () => {
  describe('コンストラクタ', () => {
    it('基本的なプロパティが正しく設定されること', () => {
      const error = new ApiError({
        kind: 'validation',
        status: 400,
        message: 'Validation error',
        errorCode: 'VALIDATION_ERROR',
        traceId: 'trace-123',
      });

      expect(error.kind).toBe('validation');
      expect(error.status).toBe(400);
      expect(error.message).toBe('Validation error');
      expect(error.errorCode).toBe('VALIDATION_ERROR');
      expect(error.traceId).toBe('trace-123');
      expect(error.name).toBe('ApiError');
    });

    it('problemDetails からフィールドエラーが抽出されること', () => {
      const error = new ApiError({
        kind: 'validation',
        status: 400,
        message: 'Validation error',
        errorCode: 'VALIDATION_ERROR',
        problemDetails: {
          error_code: 'VALIDATION_ERROR',
          title: 'Validation Error',
          status: 400,
          errors: [
            { field: 'email', message: 'Invalid email format', code: 'INVALID_EMAIL' },
            { field: 'password', message: 'Too short', code: 'TOO_SHORT' },
          ],
        },
      });

      expect(error.fieldErrors).toHaveLength(2);
      expect(error.hasFieldErrors).toBe(true);
      expect(error.getFieldError('email')).toBe('Invalid email format');
      expect(error.getFieldError('password')).toBe('Too short');
      expect(error.getFieldError('nonexistent')).toBeUndefined();
    });
  });

  describe('isRetryable', () => {
    it('リトライ可能なエラー種別で true を返すこと', () => {
      const retryableKinds: ApiErrorKind[] = ['network', 'timeout', 'server'];

      for (const kind of retryableKinds) {
        const error = new ApiError({
          kind,
          status: 500,
          message: 'Error',
          errorCode: 'ERROR',
        });
        expect(error.isRetryable).toBe(true);
      }
    });

    it('リトライ不可能なエラー種別で false を返すこと', () => {
      const nonRetryableKinds: ApiErrorKind[] = [
        'authentication',
        'authorization',
        'validation',
        'not_found',
        'conflict',
        'rate_limit',
        'unknown',
      ];

      for (const kind of nonRetryableKinds) {
        const error = new ApiError({
          kind,
          status: 400,
          message: 'Error',
          errorCode: 'ERROR',
        });
        expect(error.isRetryable).toBe(false);
      }
    });
  });

  describe('requiresAuthentication', () => {
    it('authentication エラーで true を返すこと', () => {
      const error = new ApiError({
        kind: 'authentication',
        status: 401,
        message: 'Unauthorized',
        errorCode: 'UNAUTHORIZED',
      });

      expect(error.requiresAuthentication).toBe(true);
    });

    it('他のエラー種別で false を返すこと', () => {
      const error = new ApiError({
        kind: 'authorization',
        status: 403,
        message: 'Forbidden',
        errorCode: 'FORBIDDEN',
      });

      expect(error.requiresAuthentication).toBe(false);
    });
  });

  describe('userMessage', () => {
    it('problemDetails.detail がある場合はそれを返すこと', () => {
      const error = new ApiError({
        kind: 'validation',
        status: 400,
        message: 'Validation error',
        errorCode: 'VALIDATION_ERROR',
        problemDetails: {
          error_code: 'VALIDATION_ERROR',
          title: 'Validation Error',
          status: 400,
          detail: 'Custom user message',
        },
      });

      expect(error.userMessage).toBe('Custom user message');
    });

    it('problemDetails.detail がない場合はデフォルトメッセージを返すこと', () => {
      const error = new ApiError({
        kind: 'server',
        status: 500,
        message: 'Internal error',
        errorCode: 'INTERNAL_ERROR',
      });

      expect(error.userMessage).toBeDefined();
      expect(typeof error.userMessage).toBe('string');
    });
  });

  describe('fromResponse', () => {
    it('ProblemDetails 形式のレスポンスを正しくパースすること', async () => {
      const problemDetails = {
        error_code: 'VALIDATION_ERROR',
        title: 'Validation Error',
        status: 400,
        detail: 'Input validation failed',
        trace_id: 'response-trace-123',
        errors: [{ field: 'email', message: 'Invalid' }],
      };

      const response = createMockResponse(problemDetails, {
        status: 400,
        headers: { 'content-type': 'application/problem+json' },
      });

      const error = await ApiError.fromResponse(response);

      expect(error.status).toBe(400);
      expect(error.errorCode).toBe('VALIDATION_ERROR');
      expect(error.traceId).toBe('response-trace-123');
      expect(error.fieldErrors).toHaveLength(1);
    });

    it('リクエストの traceId が優先されること', async () => {
      const response = createMockResponse(
        { error_code: 'ERROR', title: 'Error', status: 400 },
        { status: 400, headers: { 'content-type': 'application/json' } }
      );

      const error = await ApiError.fromResponse(response, 'request-trace-123');

      // レスポンスに trace_id がない場合はリクエストの traceId を使用
      expect(error.traceId).toBe('request-trace-123');
    });

    it('JSON パースに失敗してもエラーを返すこと', async () => {
      const response = {
        ok: false,
        status: 500,
        headers: {
          get: () => 'text/plain',
        },
        json: vi.fn().mockRejectedValue(new Error('Parse error')),
      } as unknown as Response;

      const error = await ApiError.fromResponse(response);

      expect(error.status).toBe(500);
      expect(error.errorCode).toBe('HTTP_500');
    });

    it('ステータスコードから正しい kind を判定すること', async () => {
      const testCases: Array<{ status: number; expectedKind: ApiErrorKind }> = [
        { status: 400, expectedKind: 'validation' },
        { status: 401, expectedKind: 'authentication' },
        { status: 403, expectedKind: 'authorization' },
        { status: 404, expectedKind: 'not_found' },
        { status: 409, expectedKind: 'conflict' },
        { status: 429, expectedKind: 'rate_limit' },
        { status: 500, expectedKind: 'server' },
        { status: 502, expectedKind: 'server' },
        { status: 503, expectedKind: 'server' },
      ];

      for (const { status, expectedKind } of testCases) {
        const response = createMockResponse(
          { error_code: 'ERROR', title: 'Error', status },
          { status, headers: { 'content-type': 'application/json' } }
        );

        const error = await ApiError.fromResponse(response);
        expect(error.kind).toBe(expectedKind);
      }
    });
  });

  describe('fromNetworkError', () => {
    it('タイムアウトエラーを正しく処理すること', () => {
      const abortError = new DOMException('Timeout', 'AbortError');

      const error = ApiError.fromNetworkError(abortError, 'trace-123');

      expect(error.kind).toBe('timeout');
      expect(error.errorCode).toBe('TIMEOUT');
      expect(error.traceId).toBe('trace-123');
    });

    it('一般的なネットワークエラーを正しく処理すること', () => {
      const networkError = new Error('Network failure');

      const error = ApiError.fromNetworkError(networkError, 'trace-123');

      expect(error.kind).toBe('network');
      expect(error.errorCode).toBe('NETWORK_ERROR');
    });

    it('Error 以外の値も処理できること', () => {
      const error = ApiError.fromNetworkError('string error');

      expect(error.kind).toBe('network');
      expect(error.errorCode).toBe('NETWORK_ERROR');
    });
  });

  describe('from', () => {
    it('既存の ApiError をそのまま返すこと', () => {
      const original = new ApiError({
        kind: 'validation',
        status: 400,
        message: 'Error',
        errorCode: 'ERROR',
      });

      const result = ApiError.from(original);

      expect(result).toBe(original);
    });

    it('Response オブジェクトを同期的に処理すること', () => {
      const response = {
        ok: false,
        status: 500,
        headers: { get: () => null },
      } as unknown as Response;

      const result = ApiError.from(response);

      expect(result.status).toBe(500);
    });

    it('その他のエラーをネットワークエラーとして処理すること', () => {
      const error = new Error('Unknown error');

      const result = ApiError.from(error);

      expect(result.kind).toBe('network');
    });
  });
});
